### ISSUE TITLE
Multiple request at same time alway throw this error - "4 DEADLINE_EXCEEDED: Deadline exceeded"

### ISSUE THREAD

**Issue by xesunny** [googleapis/nodejs-tasks#397]
#### Environment details

  - OS: Mac
  - Node.js version: 10.x
  - `@google-cloud/tasks` version: 1.7.2

#### Steps to reproduce

  1. I create a cloud task when a user perform certain activity on my app.
  2. If I get 20-30 users doing activity at the same time, I get "DEADLINE_EXCEEDED" error.

```js
if (!client) {
  let { v2beta3 } = require('@google-cloud/tasks');
  client = new v2beta3.CloudTasksClient();
  parent = client.queuePath(process.env.GOOGLE_CLOUD_PROJECT, process.env.QUEUE_LOCATION, task_queue_name);
}
client.createTask({ parent, task });
```

Thanks!

**Comment by bcoe** [comment]
@xesunny is this in a cloud function? I think you instead:

```
await client.createTask({parent, task});
```

Otherwise the cloud function retracts resources once the request is served, and you may get odd behavior.

**Comment by xesunny** [comment]
![image](https://user-images.githubusercontent.com/53507542/78852694-9f30c500-7a4f-11ea-9b09-00448921471a.png)
Here is my exact source code.

**Comment by xesunny** [comment]
> @xesunny is this in a cloud function? I think you instead:
> 
> ```
> await client.createTask({parent, task});
> ```
> 
> Otherwise the cloud function retracts resources once the request is served, and you may get odd behavior.

I'm using this code in (A) CF (B) Server side application (traditional)

I face this issue on both side. Temporary solution is : I have to submit my task sequentially. 
For example: If I have 10 tasks to be submitted, I do them one by one. 

Avg. time to submit 1 task is 200ms to 500ms. This cause unnecessary delay of ~5 seconds. 

Because of this single issue, I can't use cloud task a lot. Though I love the product.

**Comment by isaiah-coleman** [comment]
I love tasks, but having to create tasks sequentially makes it unusable for me. I need to schedule too many tasks. I've been batching promises and waiting for them to resolve, but I still get the error. Perhaps I can tweak the batch size some, but if becomes too slow I'll be forced to look for another solution :(

**Comment by yossi-eynav** [comment]
@bcoe  - We're getting this error too.
is there a limit for dispatching multiple tasks in concurrent?

**Comment by bcoe** [comment]
@yossi-eynav @isaiah-coleman I'm not sure what the hard limit is on concurrently creating tasks, but I wouldn't be shocked if one exists. My recommendation would be, if it's not reasonable to enqueue them sequentially, do something like this:

```js
const WORK_SIZE = 16; // or 32, etc.
while (work.length) {
  const chunkOfWork = work.splice(0, WORK_SIZE);
  await Promiser.all(chunkOfWork.map((chunk) => {
    return enqueueTask(chunk);
  });
}
```

☝️ this allows you to manage the number of tasks enqueued concurrently, without doing so completely sequentially.

**Comment by ChrisBeeson** [comment]
I'm getting this error dispatching a single task.

**Comment by bcoe** [comment]
@ChrisBeeson are you in a similar position as @xesunny describes, in which you enqueue a single task, but there may be multiple users causing this action concurrently.

**Comment by ChrisBeeson** [comment]
@bcoe no there are no other users

**Comment by bcoe** [comment]
@ChrisBeeson to me this sounds like a different issue than @xesunny and others are describing, which seems to be related to concurrency limits. I think it might be worth opening a new issue, so that I can help debug (I believe something else is happening to you).

**Comment by yossi-eynav** [comment]
@bcoe thanks!
I'm getting this error even when I changed the code do be sequential.
where can I get more information about this error?

**Comment by bcoe** [comment]
@yossi-eynav could you please share an example of the code you are running, and let me know what environment you're running in, i.e., Cloud Functions?

@xesunny reading your original message, the error is happening on your local OSX machine?

**Comment by yossi-eynav** [comment]
@bcoe 
my code is similar to this sample - 
https://github.com/googleapis/nodejs-tasks/blob/master/samples/createHttpTask.js.

I'm running on GKE

**Comment by ChrisBeeson** [comment]
I had the same problem, I think it's because httpRequest.body is expecting bytes, not a string. Try changing:
`    task.httpRequest.body = Buffer.from(payload).toString('base64');`
to
` task.httpRequest.body = Buffer.from(payload);`

**Comment by yossi-eynav** [comment]
@bcoe  
it's related to `@grpc/grpc-js` 
```
{ Error: 4 DEADLINE_EXCEEDED: Deadline exceeded
    at Object.callErrorFromStatus (/Volumes/dev/playground/cloud-task-test/node_modules/@grpc/grpc-js/build/src/call.js:30:26)
    at Object.onReceiveStatus (/Volumes/dev/playground/cloud-task-test/node_modules/@grpc/grpc-js/build/src/client.js:174:52)
    at Object.onReceiveStatus (/Volumes/dev/playground/cloud-task-test/node_modules/@grpc/grpc-js/build/src/client-interceptors.js:340:141)
    at Object.onReceiveStatus (/Volumes/dev/playground/cloud-task-test/node_modules/@grpc/grpc-js/build/src/client-interceptors.js:303:181)
    at Http2CallStream.outputStatus (/Volumes/dev/playground/cloud-task-test/node_modules/@grpc/grpc-js/build/src/call-stream.js:114:27)
    at Http2CallStream.maybeOutputStatus (/Volumes/dev/playground/cloud-task-test/node_modules/@grpc/grpc-js/build/src/call-stream.js:153:22)
    at Http2CallStream.endCall (/Volumes/dev/playground/cloud-task-test/node_modules/@grpc/grpc-js/build/src/call-stream.js:140:18)
    at Http2CallStream.cancelWithStatus (/Volumes/dev/playground/cloud-task-test/node_modules/@grpc/grpc-js/build/src/call-stream.js:441:14)
    at Timeout.DeadlineFilter.timer.setTimeout (/Volumes/dev/playground/cloud-task-test/node_modules/@grpc/grpc-js/build/src/deadline-filter.js:59:28)
    at ontimeout (timers.js:436:11)
```

**Comment by isaiah-coleman** [comment]
@yossi-eynav if you look, a lot of '@google-cloud' libraries with a '@grpc/grpc-js' dependency have an open issue related to concurrency. The only thing I've read that helps people seems to be upgrading your Node version to 'v13'. However, that hasn't fixed the issue for me. 

see: https://github.com/grpc/grpc-node/issues/1158 it might help

**Comment by yossi-eynav** [comment]
@bcoe @isaiah-coleman 
I've set the flag `fallback` to be true and now I have 0 errors on creating 1000 tasks in concurrent..

it seems that the problem is with `gax-nodejs` library -> can you share the changes between the regular flow and the fallback flow since I'm not familiar with this library at all?

**Comment by isaiah-coleman** [comment]
@yossi-eynav I'm not sure I can help you. I'm not a contributor on the library, merely just sharing what seems to have helped others. Can you explain how you used the `fallback` flag to fix the errors? It might be a solution I can use.

**Comment by alexander-fenster** [comment]
@yossi-eynav Setting `fallback` to true enables a different transport (the one that was initially supposed for browsers) - instead of using [gRPC](https://grpc.io/), it serializes your requests and sends them over regular HTTP/1 connection with `node-fetch` to a different endpoint.  When you enable fallback, you don't do any gRPC requests at all - it uses totally different stack. That might serve as a good workaround, but we'd like to investigate what's going on.

Would it be possible for you to measure the time - how long does it take for it to fail with `DEADLINE EXCEEDED`?

**Comment by xesunny** [comment]
> @yossi-eynav could you please share an example of the code you are running, and let me know what environment you're running in, i.e., Cloud Functions?
> 
> @xesunny reading your original message, the error is happening on your local OSX machine?

Yes, It's happening on my Mac OSX machine but I have got this error on "Google Cloud instance" & "Google Cloud run" as well.

**Comment by xesunny** [comment]
> @yossi-eynav Setting `fallback` to true enables a different transport (the one that was initially supposed for browsers) - instead of using [gRPC](https://grpc.io/), it serializes your requests and sends them over regular HTTP/1 connection with `node-fetch` to a different endpoint. When you enable fallback, you don't do any gRPC requests at all - it uses totally different stack. That might serve as a good workaround, but we'd like to investigate what's going on.
> 
> Would it be possible for you to measure the time - how long does it take for it to fail with `DEADLINE EXCEEDED`?

Can you give an example on how to use "fallback: true" ?  I don't have much experience with gprc.

**Comment by yossi-eynav** [comment]
@alexander-fenster 
~120s (created 1000 tasks in parallel)

**Comment by yossi-eynav** [comment]
@xesunny 
```
const {CloudTasksClient} = require('@google-cloud/tasks');
const client = new CloudTasksClient({ fallback: true })
```

**Comment by isaiah-coleman** [comment]
Are the requests over HTTP/1 regional?  Electing for { falback: true } gives me the following

 `error. {"code":3,"message":"Location 'us-east1' is not a valid location. Use ListLocations to list valid locations.","details":[]}
`

I know this is the correct region, when i switch out the region I receive the following:

`{"code":3,"message":"Location must equal us-east1 because the App Engine app that is associated with this project is located in us-east1","details":[]}`

For due diligence this is me checking the ListLocations 

> $ gcloud tasks locations list
> NAME      FULL_NAME
> us-east1  projects/[project-name]/locations/us-east1

I'm grateful for any help that can be provided.

**Comment by yossi-eynav** [comment]
@isaiah-coleman  i'm on us-center1

**Comment by xesunny** [comment]
am also getting same error :
Error: {"code":3,"message":"Location 'asia-south1' is not a valid location. Use ListLocations to list valid locations.","details":[]}

**Comment by yossi-eynav** [comment]
@alexander-fenster @xesunny 
any update on this?

**Comment by alexander-fenster** [comment]
@yossi-eynav Not yet, it's on my list, please expect an update soon.

**Comment by xesunny** [comment]
Thanks team. Really looking forward for it.

**Comment by Iamkhova** [comment]
I encountered this problem also and the workaround mentioned above worked. 
`const client = new CloudTasksClient({ fallback: true })`

Though still interested in what is originally causing this error. I only noticed it in the last few days. 

`app:///../node_modules/@google-cloud/tasks/node_modules/@grpc/grpc-js/build/src/call.js in Object.callErrorFromStatus at line 30:26
 * Construct a ServiceError from a StatusObject. This function exists primarily
 * as an attempt to make the error stack trace clearly communicate that the
 * error is not necessarily a problem in gRPC itself.
 * @param status
 */
function callErrorFromStatus(status) {
    const message = `${status.code} ${constants_1.Status[status.code]}: ${status.details}`;
    return Object.assign(new Error(message), status);
}
exports.callErrorFromStatus = callErrorFromStatus;
class ClientUnaryCallImpl extends events_1.EventEmitter {
    constructor() {
        super();
    }
    cancel() {`

**Comment by shayne-lp** [comment]
Hi there, I too am curious if there are any workarounds for `us-east1`, as the fallback approach seemingly cannot be used there. Thanks!

**Comment by shayne-lp** [comment]
Hi there, just wanted to circle back and see if anyone has made any progress on the above. In particular, how to use the fallback option in us-east1. Thanks!

**Comment by shayne-lp** [comment]
Hey @alexander-fenster are there any updates you can provide for non us-central users? Google support punted us back here :( this is really biting us in prod. Thanks

**Comment by alexander-fenster** [comment]
@shayne-lp Yes I'm working with support folks as well investigating what's going on here. Please expect to have an update later today or tomorrow. I'm sorry that it affects you folks. There's a lot of moving things there between the client library and the backend, I'l trying to understand what's going on.

**Comment by gijo-varghese** [comment]
I almost had the same error. In my case, I was not awaiting the call. Adding `await` or `Promise.all()` fixed the issue for me. Here is the detailed SO question: [DEADLINE_EXCEEDED while adding tasks to Google Cloud Tasks](https://stackoverflow.com/questions/61822332/deadline-exceeded-while-adding-tasks-to-google-cloud-tasks)

**Comment by alexander-fenster** [comment]
Wow thank you @gijo-varghese! Can you folks please check if this applies to your code?

**Comment by shayne-lp** [comment]
Hi @alexander-fenster, thanks for getting back to us. And thank you for your reply too, @gijo-varghese. We are indeed already using `Promise.all()`. After much trial and error, the issue we are seeing is the **number** of promises you are attempting to resolve dictates if you get the `DEADLINE_EXCEEDED` error or not. For example, if you wanted to do 10,000 promises, or even 1,000 promises in bulk, we would get the error. If we do batches of 100, we do not. The other problem is we are unable to catch this error and retry any that failed (it just blows us up).

Any guidance on the upper bounds there (whether dictated by this library, or Google Tasks itself) would be appreciated! Thanks all

**Comment by xesunny** [comment]
Still facing the issue.

**Comment by gbhojak** [comment]
I am seeing this issue as well even when I try to queue a single request with a Google Cloud Run service.
This is what I am doing:

`   
 const request = {parent, task};

    try {
        const [response] = await cloudTasksClient.createTask(request);
        console.error('Queued the task:', response.name);
    } catch (e) {
        console.error('Error queuing the task', e);
    }`
I am using "@google-cloud/tasks": "^2.0.0".

**Comment by isaiah-coleman** [comment]
Hi, I just wanted to circle back with an update. We tuned the number of promises in a batch down to `100` and used `Promise.all()` to wait for them to resolve. This quieted the error for a time, but we experienced the issue again today. There have been no changes to code or infrastructure. 

We can continue to fine-tune batch sizes, but without any guidance it's hard to know if it will fix the issue long term. Thanks

**Comment by bcoe** [comment]
@isaiah-coleman @xesunny sorry that its been a little while since an update, we've been discussing this issue internally a bit.

Here's where I end up being a bit confused:

---

I've created a task queue as follows `gcloud tasks queues create echo-function`. I can then run the following code, and have yet to see a DEADLINE_EXCEEDED issue:

```js
const {CloudTasksClient} = require('@google-cloud/tasks')

let client;
async function performTask () {
  if (!client) {
    client = new CloudTasksClient()
  }
  const queuePath = client.queuePath('long-door-651', 'us-central1', 'echo-function');
  const [resp] = await client.createTask({
    parent: queuePath,
    task: {
      httpRequest: {
        httpMethod: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        url: 'https://us-central1-long-door-651.cloudfunctions.net/echo-function',
        body: Buffer.from(JSON.stringify({hello: 'world'})),
      },
    },
  });
  console.info('we finished creating task', resp);
}

for (let i = 0; i < 2048; i++) {
  performTask()
}
```

☝️ note that this is creating thousands of tasks concurrently.

---

Here are the exact versions of the dependencies I have installed:

```bash
bencoe@bencoe:~/google/tasks-perf$ npm ls
tasks-perf@1.0.0 /usr/local/google/home/bencoe/google/tasks-perf
└─┬ @google-cloud/tasks@2.1.0
  └─┬ google-gax@2.6.3
    ├─┬ @grpc/grpc-js@1.1.2
    │ └── semver@6.3.0 deduped
    ├─┬ @grpc/proto-loader@0.5.5
    │ ├── lodash.camelcase@4.3.0
    │ └── protobufjs@6.10.1 deduped
    ├── @types/long@4.0.1
    ├─┬ abort-controller@3.0.0
    │ └── event-target-shim@5.0.1
    ├─┬ duplexify@3.7.1
    │ ├─┬ end-of-stream@1.4.4
    │ │ └─┬ once@1.4.0
    │ │   └── wrappy@1.0.2
    │ ├── inherits@2.0.4
    │ ├─┬ readable-stream@2.3.7
    │ │ ├── core-util-is@1.0.2
    │ │ ├── inherits@2.0.4 deduped
    │ │ ├── isarray@1.0.0
    │ │ ├── process-nextick-args@2.0.1
    │ │ ├── safe-buffer@5.1.2
    │ │ ├─┬ string_decoder@1.1.1
    │ │ │ └── safe-buffer@5.1.2 deduped
    │ │ └── util-deprecate@1.0.2
    │ └── stream-shift@1.0.1
    ├─┬ google-auth-library@6.0.5
    │ ├── arrify@2.0.1
    │ ├── base64-js@1.3.1
    │ ├─┬ ecdsa-sig-formatter@1.0.11
    │ │ └── safe-buffer@5.1.2 deduped
    │ ├── fast-text-encoding@1.0.3
    │ ├─┬ gaxios@3.0.4
    │ │ ├── abort-controller@3.0.0 deduped
    │ │ ├── extend@3.0.2
    │ │ ├─┬ https-proxy-agent@5.0.0
    │ │ │ ├─┬ agent-base@6.0.1
    │ │ │ │ └── debug@4.1.1 deduped
    │ │ │ └── debug@4.1.1 deduped
    │ │ ├── is-stream@2.0.0
    │ │ └── node-fetch@2.6.0 deduped
    │ ├─┬ gcp-metadata@4.1.4
    │ │ ├── gaxios@3.0.4 deduped
    │ │ └─┬ json-bigint@1.0.0
    │ │   └── bignumber.js@9.0.0
    │ ├─┬ gtoken@5.0.2
    │ │ ├── gaxios@3.0.4 deduped
    │ │ ├─┬ google-p12-pem@3.0.2
    │ │ │ └── node-forge@0.9.1
    │ │ ├── jws@4.0.0 deduped
    │ │ └── mime@2.4.6
    │ ├─┬ jws@4.0.0
    │ │ ├─┬ jwa@2.0.0
    │ │ │ ├── buffer-equal-constant-time@1.0.1
    │ │ │ ├── ecdsa-sig-formatter@1.0.11 deduped
    │ │ │ └── safe-buffer@5.1.2 deduped
    │ │ └── safe-buffer@5.1.2 deduped
    │ └─┬ lru-cache@6.0.0
    │   └── yallist@4.0.0
    ├── is-stream-ended@0.1.4
    ├── lodash.at@4.6.0
    ├── lodash.has@4.5.2
    ├── node-fetch@2.6.0
    ├─┬ protobufjs@6.10.1
    │ ├── @protobufjs/aspromise@1.1.2
    │ ├── @protobufjs/base64@1.1.2
    │ ├── @protobufjs/codegen@2.0.4
    │ ├── @protobufjs/eventemitter@1.1.0
    │ ├─┬ @protobufjs/fetch@1.1.0
    │ │ ├── @protobufjs/aspromise@1.1.2 deduped
    │ │ └── @protobufjs/inquire@1.1.0 deduped
    │ ├── @protobufjs/float@1.0.2
    │ ├── @protobufjs/inquire@1.1.0
    │ ├── @protobufjs/path@1.1.2
    │ ├── @protobufjs/pool@1.1.0
    │ ├── @protobufjs/utf8@1.1.0
    │ ├── @types/long@4.0.1 deduped
    │ ├── @types/node@13.13.14
    │ └── long@4.0.0
    ├─┬ retry-request@4.1.1
    │ ├─┬ debug@4.1.1
    │ │ └── ms@2.1.2
    │ └─┬ through2@3.0.2
    │   ├── inherits@2.0.4 deduped
    │   └── readable-stream@2.3.7 deduped
    ├── semver@6.3.0
    └── walkdir@0.4.1
```

And my Node.js runtime `v14.5.0`.

----

@isaiah-coleman @gbhojak @xesunny could you provide me [a gist](https://gist.github.com/), with this same information:

1. the code that is giving you DEADLINE_EXCEEDED (_with secrets redacted_).
2. the runtime that you're running in.
3. your `npm ls` output (_if you're running locally_).

**Comment by isaiah-coleman** [comment]
Thanks @bcoe I understand your confusion. The error behavior seems to be non-deterministic. You would have to run the snippet every day for weeks to have the error manifest. We run the gist below daily in sets a couple times a day. 
1. the code that is giving you DEADLINE_EXCEEDED : [gist](https://gist.github.com/isaiah-coleman/c1439728ddbe654fb03d9f81ecc7188d)
2. the runtime that you're running in. Node `v10.21.1`
3. npm ls: [npm ls](https://gist.github.com/isaiah-coleman/7dfee4c7e859c6a92adce0c20a255c52)

**Comment by xesunny** [comment]
Thanks @bcoe , can you change the location of your queue from us-central1 to asia-northeast1 & then compare the result?

**Comment by bcoe** [comment]
@isaiah-coleman, @xesunny, thank you for continuing to help me dig into this 👍 

@xesunny I will try @asia-northeast, is your Cloud Function and/or server also located near this data center?

@isaiah-coleman what location are you in?

**Comment by isaiah-coleman** [comment]
@bcoe we're using `us-east1`

**Comment by bcoe** [comment]
@isaiah-coleman thank you for the reproduction 👍 (_and I'm sorry it took me 8 days to respond_).

_One more point of clarification, in what environment are you running the script? (it sounds like you're not in a GCP environment, as in @xesunny's original report which was on Mac OS)._

Once I've confirmed this, I think I have a plan for some folks to bring this issue up with internally.

**Comment by isaiah-coleman** [comment]
Thanks for the response @bcoe, we're actually using Cloud Run.

**Comment by smaeda-ks** [comment]
@bcoe 
I also got this same error today and found this topic interesting.

I'm running my Node.js app on a single GAE (flexible) instance and creating tasks from there. The code is really simple, and I'm _not_ creating tasks concurrently but rather creating sequentially within a `while() { .. }`.

```js
    while (some_conditions) {
        [ ... ]
        const [response] = await cloudtask.createTask({parent: parent, task: task});
        const name = response.name;
        console.log(`Created task ${name}`);
    }
```

<img width="1124" alt="Screen Shot 2020-07-31 at 8 08 31" src="https://user-images.githubusercontent.com/11495867/88983861-1512b200-d307-11ea-8812-ff8e565c1c3d.png">

I don't think I can easily reproduce this at all as this is the first time I see this error. I've never seen this before in the last few weeks (since I started using Tasks).

GAE location: `us-central1`
GAE runtime: `nodejs12`
@google-cloud/tasks version: `2.0.1`

**Comment by smaeda-ks** [comment]
Also, in my case above, it seems there was roughly about ~10 seconds stuck time before it raises DEADLINE_EXCEEDED error. I could confirm this from the GAE logging timestamp:

> 2020-07-31 07:30:33.000 JST Queue id: 5690167314612224 created successfully.
> 2020-07-31 07:30:43.000 JST (node:25) UnhandledPromiseRejectionWarning: Error: 4 DEADLINE_EXCEEDED: Deadline exceeded at...

Note, the log message on `2020-07-31 07:30:33.000 JST` indicates we successfully created a new Datastore entity. Since we use Express we immediately return `next()` after this `consolg.log()` line, and move to that function where we create tasks. So it is accurate to say ~10 seconds, I think.

@bcoe @alexander-fenster

**Comment by bcoe** [comment]
@smaeda-ks, @isaiah-coleman,  @xesunny,  (and other folks in this thread bumping into this issue).

We've opened an internal tracking thread, if would be helpful if people could send me two pieces of  information:

1. your GCP project id.
2. a timestamp of from when the `DEADLINE_EXCEEDED` error was observed.

If you don't feel comfortable sharing in this thread, you can email `bencoe @ google.com`.

**Comment by ehsan-karamad** [comment]
Has anyone tried setting the deadline explicitly? Perhaps to longer than `10s.`

For instance
`
const request = {
  parent: queue_path,
  task: task,
}

const call_options = {
    timeout: 100
}

const response = await client.createTask(request, call_options);
`

**Comment by bcoe** [comment]
@smaeda-ks, @isaiah-coleman, @xesunny, there is an internal ticket now, and we're investigating some potential root causes. 

However, we would  be interested to see if extending the timeout shields you from this problem. So, if you have the opportunity, do try @ehsan-karamad's recommendation of a higher timeout:

```js
const request = {
  parent: queue_path,
  task: task,
}

const call_options = {
    timeout: 20000 // millis
}

const response = await client.createTask(request, call_options);
```

**Comment by smaeda-ks** [comment]
@bcoe is the default timeout `10000`?
https://github.com/googleapis/nodejs-tasks/blob/7dda2873c93295ccf19c6dc66daf4625cbe115bb/src/v2/cloud_tasks_client_config.json#L43

then my case was likely hitting that:
https://github.com/googleapis/nodejs-tasks/issues/397#issuecomment-666862807

**Comment by smaeda-ks** [comment]
though I'm a bit skeptical that increasing the timeout limit will actually mitigate the issue if the default is 10 seconds. That seems already long enough.

**Comment by shayne-lp** [comment]
One would certainly hope 10 seconds is long enough. Tuning magic constants
to address nondeterministic bugs almost never feels like the right
solution, and just masks the underlying root cause. But if anyone finds it
helps them, please do let the rest of us affected masses know. :)

I hope this long-standing issue gets the attention and priority it deserves
within Google. Thanks all for the continued effort!

**Comment by banikKGP** [comment]
> @isaiah-coleman i'm on us-center1

I am also facing the same error but it worked with this. Can you please explain why?

**Comment by richardarpino** [comment]
I too was using the sample code and ran into this issue. No loops - just a single request.

I am running on Node v10.17.0 using the v2.1.0 of the library on Debian.

The increase in timeout helped me a little - it showed an initial IAM error. I am using a service account which needed the Service Account Role (See https://github.com/googleapis/nodejs-tasks/issues/249#issuecomment-525827441)

That got a response but it is consistently over 10 seconds, not by much but enough to hit the deadline.

What I noticed was the the initial createTask is way slower than subsequent ones, which execute almost immediately.

**Comment by liorschwimmer** [comment]
I got the same issue but with '@google-cloud/secret-manager'
calling to smclient.accessSecretVersion  returns error: Error: 4 DEADLINE_EXCEEDED: Deadline exceeded.
The error is not consistent. 

I posted my code here:

[https://stackoverflow.com/q/64751386/3275324](https://stackoverflow.com/q/64751386/3275324)

**Comment by marcosrubio3** [comment]
Any updates on this?

I am getting the same error while creating a task inside a cloud function. Same as @richardarpino, no loops, just a single request.

**Comment by bcoe** [comment]
@marcosrubio3 👋  we're having some internal conversations about a potentially related issue, I will keep this thread updated.

**Comment by grangerjs** [comment]
Watching here as I am seeing the exact same issue that @liorschwimmer saw with `@google-cloud/secret-manager` on a Fastify app in Cloud Run. It only seems to fail every couple of days or so. I can't reproduce it locally. 

Cloud Run Log:
![Pasted_Image_2_16_21__4_13_PM](https://user-images.githubusercontent.com/42156515/108133616-196ba600-7072-11eb-8865-2ac11d149e7c.png)

Implementation:
![Pasted_Image_2_16_21__4_17_PM](https://user-images.githubusercontent.com/42156515/108133920-926afd80-7072-11eb-9e5b-984fdc58737a.png)

Could also be user error... haven't ruled that out yet.

**Comment by dhara15101991** [comment]
Has anyone solved the issue? I facing the same issue. I need to create 500-600 tasks when a user added. I am using a loop to create it. But facing an error.

**Comment by sofisl** [comment]
Internal tracking number: 186681285

**Comment by superzadeh** [comment]
We had the same issue come up in production in the past 2-4 days. Any updates on this? This is affecting us in production.

**Comment by bcoe** [comment]
@shayne-lp @superzadeh, since @sofisl opened the ticket, it has been assigned to an engineer and is being investigated.

I'm sorry about how long this issue has been affecting folks, and will keep sharing information with the product team.

**Comment by superzadeh** [comment]
Any updates @bcoe or @sofisl? or a workaround maybe? This is really starting to become a problem in our production environment

**Comment by nikkanetiya** [comment]
@bcoe Appreciate if we can get any quick help on this one.

We are using GCP heavily for our infrastructure. Currently, in a new microservice hosted on GCP Kubernetes, we are getting this error so many times. Earlier, it was coming for firestore, google cloud storage, and task requests, we tried everything updated all lib. version, node version to 14.x.

But currently, we are seeing this error for the task request these are a number for this service.

Cloud Task got created successfully:- 131,148
Cloud Task got error:- 3,504

In case you need anything from our side let me know I can provide many examples, traces, and project ID if required. This is the only error preventing us right now to enable this feature for everyone and affecting us seriously on production.

We have P1 support as well but the reason to post it here is in case I can provide some quick trace or detail if you guys need anything from us check this in detail.


<img width="1519" alt="Screenshot 2021-05-19 at 2 20 39 PM" src="https://user-images.githubusercontent.com/6582898/118784167-66e3f300-b8ad-11eb-82a6-ff5228c769cc.png">

**Comment by superzadeh** [comment]
Very similar symptoms to you @nikkanetiya and unreliability of most services we use on GCP that rely on GRPC (so, most of them actually). 

We also ran into https://issuetracker.google.com/issues/158014637?pli=1 last year (which is still unresolved!) and already spent an unnecessary amount of time migrating our code from a serverless architecture to AppEngine (which was the only option we had with the time we had). I have the impression that the GRPC/core libs that are used by GCP services are fundamentally broken.

We already spent too much time trying to find workarounds, and the fact that these issues stay open for over a year made us take the decision to move away all our stuff back to AWS (we used it for most our product, and tried GCP/Firebase for a new mobile product ==> we'll migrate all that to AWS). It is a significant time investment (again), coming at the worse timing possible (we're fundraising in 3 months), and I would have really preferred to find another resolution.

This stuff just doesn't happen on AWS (not affiliated to them, can't say for Azure or others providers but we're having a *much* better experience with AWS), and even if it does, you get daily updates on the status of the issue/investigation, and resolution happens much faster.

I am mostly sharing this here for transparency towards the community who might have to make hard decisions in order to deliver the bare minimum of UX to their users by providing a reliable infrastructure for their services.

**Comment by GlennChon** [comment]
Might be helpful for some.  
My setup:  cloud function as a webhook to schedule multiple tasks that call a new cloud function webhook with dynamic parameters

I had this issue with creating multiple tasks in the queue and it looks like I wasn't handling the promises properly.

`
import {createTask} from 'wherever/your/cloud-task-function/is)

let arr = [
let promises = []
arr.forEach((item)=>{
     const {prop1, prop2, propN...} = item
     promises.push(createTask({prop1, prop2, propN...))
}
return Promise.all(promises).then(()=>{
     return res.status(201).send('Task Generator Initialized')
}).catch((err)=>{
     return res.status(500).send(err.message)
}
`

**Comment by bcoe** [comment]
@nikkanetiya @superzadeh are you scheduling many tasks concurrently, if you perform `Promise.all` with a large number of tasks, you can run into contention on resources, which can lead to issues like `DEADLINE_EXCEEDED`.

If you have many tasks to enqueue in on job, an approach I take is as follows:

```
const tasksToCreate = [...]; // an array with a large number of tasks to enqueue.
const WORK_SIZE = 32; // some size of work to perform.
while (work.length > 0) {
  const work = tasksToCreate.slice(0, WORK_SIZE).map(() => {
     return createTaskPromise();
   });
  await Promise.all(work);
}
```

**Comment by bcoe** [comment]
@xesunny I believe this approach, in my last comment, of chunking up work will likely work for your use case too.

You do not need to perform one task at a time, but you should pick some upper bound on the amount of work that is processed concurrently, and you should await completion, otherwise you can end up with an unhandled rejection.

**Comment by superzadeh** [comment]
> @nikkanetiya @superzadeh are you scheduling many tasks concurrently, if you perform `Promise.all` with a large number of tasks, you can run into contention on resources, which can lead to issues like `DEADLINE_EXCEEDED`.
> 
> 
> 
> If you have many tasks to enqueue in on job, an approach I take is as follows:
> 
> 
> 
> ```javascript
> 
> const tasksToCreate = [...]; // an array with a large number of tasks to enqueue.
> 
> const WORK_SIZE = 32; // some size of work to perform.
> 
> while (work.length > 0) {
> 
>   const work = tasksToCreate.slice(0, WORK_SIZE).map(() => {
> 
>      return createTaskPromise();
> 
>    });
> 
>   await Promise.all(work);
> 
> }
> 
> ```

Thanks for the reply 🙏 That is what we're doing, except we use batch sizes of 100. I'll try to tune it down to 32, but overall this feels pretty limited if even a batch size of 100 cause contention on the system. What I've typically found when using similar queueing service is the ability to queue batches: the client sends X amount of tasks to queue, and the contention/batch queue is managed by the system itself (in this case, Cloud Task). This feels like a critical feature missing that is now being thrown at users to "deal with it".

How do you recommend scaling this approach? We have larger batches (currently up to 20k, but we plan to scale it to at least 10x, maybe 50x towards December this year), and queuing them "32 by 32" is simply going to be slow. We are currently queueing the tasks through a fan-out Cloud Scheduler job, calling AppEngine which is queueing the requests. I fear that with a batch size of 32, the AppEngine request will simply timeout before finishing to queue all the tasks.

Any ideas?

**Comment by nikkanetiya** [comment]
> @nikkanetiya @superzadeh are you scheduling many tasks concurrently, if you perform `Promise.all` with a large number of tasks, you can run into contention on resources, which can lead to issues like `DEADLINE_EXCEEDED`.
> 
> If you have many tasks to enqueue in on job, an approach I take is as follows:
> 
> ```js
> const tasksToCreate = [...]; // an array with a large number of tasks to enqueue.
> const WORK_SIZE = 32; // some size of work to perform.
> while (work.length > 0) {
>   const work = tasksToCreate.slice(0, WORK_SIZE).map(() => {
>      return createTaskPromise();
>    });
>   await Promise.all(work);
> }
> ```

@bcoe Thank you for your reply.

We are using this with worker listening to the messages from the PubSub queue, so we are not always adding 100s task queue at the same time concurrently so we will not able to chuck it and honestly, I think currently a single pod is only processing 100 messages at a time (not all message will create up the create task).

And Just to give you some more context let's say if we will always get `DEADLINE_EXCEEDED` error on failer, we can still add some backoff logic to retry it again, But the real problem here is on many tasks are actually getting created on Cloud Task even though we got `DEADLINE_EXCEEDED` so we have no idea on our end whatever that task is created or not and will it get executed at that or not.

**Comment by bcoe** [comment]
@nikkanetiya @superzadeh have you tried tuning timeout in call options;

```javascript
const call_options = {
    timeout: 200000 // millis
}
const response = await client.createTask(request, call_options);
```

I believe this can be overridden on a request by request basis, and combined with chunking up your work, you may find that you avoid the DEADLINE_EXCEEDED issue.

**Comment by smaeda-ks** [comment]
https://github.com/googleapis/nodejs-tasks/blob/v2.3.1/src/v2beta2/cloud_tasks_client.ts#L3006-L3022

I've been calling this method explicitly after task creation (e.g., using `Promise.all()`) and so far not seeing this DEADLINE_EXCEEDED error as I used to be.

**Comment by superzadeh** [comment]
> @nikkanetiya @superzadeh have you tried tuning timeout in call options;
> 
> ```js
> const call_options = {
>     timeout: 200000 // millis
> }
> const response = await client.createTask(request, call_options);
> ```
> 
> I believe this can be overridden on a request by request basis, and combined with chunking up your work, you may find that you avoid the DEADLINE_EXCEEDED issue.

I will give this a try @bcoe, but this raises even more concerns regarding scalability: if we are only creating tasks 32 at a time, with a timeout of 200 000ms, how are we supposed to scale this approach? We currently have batches of 20k tasks, and plan to grow that to at least 100k within a few months. What is your recommendation here?

Also, if you have any status updates on the internal tracking number `186681285`, it would go a long way to create transparency and help build trust in GCP's ability to deliver a reliable platform (is it sitting somewhere in a backlog, in progress, stuck without being able to reproduce, found the issue and working on a fix?)

**Comment by superzadeh** [comment]
> > @nikkanetiya @superzadeh have you tried tuning timeout in call options;
> > ```js
> > const call_options = {
> >     timeout: 200000 // millis
> > }
> > const response = await client.createTask(request, call_options);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I believe this can be overridden on a request by request basis, and combined with chunking up your work, you may find that you avoid the DEADLINE_EXCEEDED issue.
> 
> I will give this a try @bcoe, but this raises even more concerns regarding scalability: if we are only creating tasks 32 at a time, with a timeout of up to `200000` ms, how are we supposed to scale this approach? We currently have batches of 20k tasks, and plan to grow that to at least 100k within a few months. What is your recommendation here?
> 
> Also, if you have any status updates on the internal tracking number `186681285`, it would go a long way to create transparency and help build trust in GCP's ability to deliver a reliable platform (is it sitting somewhere in a backlog, in progress, stuck without being able to reproduce, found the issue and working on a fix?)

We gave this a shot @bcoe, and when we add the `call_options` as you have suggested, all the tasks that we create turn into `POST`, instead of `GET` 🤷‍♂️. 
I assume that passing callOptions overrides the method we define on the task, I tried to look for some documentation for the call options, but couldn't find any, except this open issue: https://github.com/googleapis/nodejs-tasks/issues/320. Because of this, we couldn't try that in an environment where we have enough scale to see the `DEADLINE_EXCEEDED` issue.

Here's how we create the tasks, do you see any reason why they end up being `POST` instead of `GET`?

```ts

constructor() {
    // We only create the task client once
    const client = new CloudTasksClient();
    const project = process.env.GCLOUD_PROJECT!;
    const location = process.env.QUEUE_LOCATION;
    this.taskClient = client
}    
 
 
public async queueTask = (uri: string, queueName: string) => { 
    // we reuse the task client
    const parent = this.taskClient.queuePath(project, location, queueName);

    const task: protos.google.cloud.tasks.v2.ITask = {
      appEngineHttpRequest: {
        httpMethod: "GET",
        relativeUri: uri,
      },
    };

    // Send create task request.
    const request: protos.google.cloud.tasks.v2.ICreateTaskRequest = {
      parent,
      task,
    };
    
    // Creating the task like this results in the task sent to our AppEngine as POST
    const [response] = await this.taskClient.createTask(request, callOptions)
    
     // Creating the task like this results in the task sent to our AppEngine as GET
    const [response] = await this.taskClient.createTask(request)
    
    //...
}
```

**Comment by superzadeh** [comment]
I'm really sorry if I come off as being pushy, but we're still having this error in production. Any updates you can share? My leadership team is going to pull the gun and start a migration to AWS if I come back to them saying there's still no updates on this.

**Comment by bcoe** [comment]
@superzadeh `call_options` should have no affect on whether the event dispatched to you by Cloud Tasks is ultimately a `POST` or ` `GET`, could you please open a separate issue describing this?

**Comment by bcoe** [comment]
@xesunny I'm doing some cleanup up issues, and am closing this as I haven't heard back from you in a while. 

If folks are continuing to bump into problems, please don't hesitate to open a new issue, including a code snippet, and your gcloud project ID.

**Comment by bcoe** [comment]
@shayne-lp if you open a new issue with a code snippet that demonstrates the issue you're bumping into, along with your project ID, we can share it with the internal engineering team.

If you don't feel comfortable sharing a project ID on GitHub, you can send an email to `bencoe @ google.com`.

**Comment by jorgemgr94** [comment]
People who are still experiencing this issue consider the workaround mentioned by @xesunny .
Currently, we're enqueuing batches of 500 tasks concurrently without any problem 👍 

> @xesunny
> 
> ```
> const {CloudTasksClient} = require('@google-cloud/tasks');
> const client = new CloudTasksClient({ fallback: true })
> ```
